lib/wallet/src/Cardano/Wallet/Flavor.hs

```diff
@@ -23,18 +31,22 @@ import Cardano.Wallet.Primitive.AddressDiscovery.Shared
     ( SharedState (..) )
 
 data WalletFlavorS s n k  where
-    ShelleyWallet :: WalletFlavorS (SeqState n k) n k
-    ByronWallet :: WalletFlavorS (RndState n) n k
-    SharedWallet :: WalletFlavorS (SharedState n k) n k
+    ShelleyWallet :: WalletFlavorS (SeqState n ShelleyKey) n ShelleyKey
+    IcarusWallet :: WalletFlavorS (SeqState n IcarusKey) n IcarusKey
+    ByronWallet :: WalletFlavorS (RndState n) n ByronKey
+    SharedWallet :: WalletFlavorS (SharedState n k) n SharedKey
 
 class WalletFlavor s n k where
     walletFlavor :: WalletFlavorS s n k
 
-instance WalletFlavor (SeqState n k) n k where
+instance WalletFlavor (SeqState n IcarusKey) n IcarusKey where
+    walletFlavor = IcarusWallet
+
+instance WalletFlavor (SeqState n ShelleyKey) n ShelleyKey where
     walletFlavor = ShelleyWallet
 
-instance WalletFlavor (RndState n) n k where
+instance WalletFlavor (RndState n) n ByronKey where
```

Comment:
👍 I think that this also makes it possible to remove the `AllowAmbiguousTypes` extension?